state/indexer/query_range.go

```diff
@@ -0,0 +1,123 @@
+package indexer
+
+import (
+	"time"
+
+	"github.com/tendermint/tendermint/libs/pubsub/query"
+)
+
+// QueryRanges defines a mapping between a composite event key and a QueryRange.
+//
+// e.g.account.number => queryRange{lowerBound: 1, upperBound: 5}
+type QueryRanges map[string]QueryRange
+
+// QueryRange defines a range within a query condition.
+type QueryRange struct {
+	LowerBound        interface{} // int || time.Time
+	UpperBound        interface{} // int || time.Time
+	Key               string
+	IncludeLowerBound bool
+	IncludeUpperBound bool
+}
+
+// AnyBound returns either the lower bound if non-nil, otherwise the upper bound.
+func (qr QueryRange) AnyBound() interface{} {
+	if qr.LowerBound != nil {
+		return qr.LowerBound
+	}
+
+	return qr.UpperBound
+}
+
+// LowerBoundValue returns the value for the lower bound. If the lower bound is
+// nil, nil will be returned.
+func (qr QueryRange) LowerBoundValue() interface{} {
+	if qr.LowerBound == nil {
+		return nil
+	}
+
+	if qr.IncludeLowerBound {
+		return qr.LowerBound
+	}
+
+	switch t := qr.LowerBound.(type) {
+	case int64:
+		return t + 1
+
+	case time.Time:
+		return t.Unix() + 1
+
+	default:
+		panic("not implemented")
```

Comment:
should we return an error? or avoid needing to panic in some other way (by like, I presume having a validate method on QueryRange?)

Reply:
This was a drop-n-lift from existing code. Things you find odd in this, let's put them in an issue and address them?

Reply:
Sounds good to me!